packages/main/src/plugin/kubernetes-client.ts

```diff
@@ -1049,64 +1010,157 @@ export class KubernetesClient {
-    let telemetryOptions = {};
+  async createResources(context: string, manifests: any[], namespace?: string): Promise<void> {
+    await this.syncResources(context, manifests, 'create', namespace);
+  }
+
+  /**
+   * Apply a given yaml file to a context, i.e. 'kubectl apply -f'. Resources that exist
+   * on the context are patched, and any new resources are created.
+   *
+   * @param context a context
+   * @param filePath file system path to a YAML Kubernetes spec
+   * @param namespace the namespace to use for any resources that don't include one
+   * @return an array of resources created
+   */
+  async applyResourcesFromFile(context: string, filePath: string, namespace?: string): Promise<KubernetesObject[]> {
+    const manifests = await this.loadManifestsFromFile(filePath);
+    return this.syncResources(context, manifests, 'apply', namespace);
+  }
+
+  /**
+   * Apply a given yaml file to a context, i.e. 'kubectl apply -f'. Resources that exist
+   * on the context are patched, and any new resources are created.
+   *
+   * @param context a context
+   * @param filePath file system path to a YAML Kubernetes spec
+   * @param namespace the namespace to use for any resources that don't include one
+   * @return an array of resources created
+   */
+  async applyResources(context: string, manifests: unknown[], namespace?: string): Promise<KubernetesObject[]> {
+    return this.syncResources(context, manifests, 'apply', namespace);
+  }
+
+  /**
+   * Applies or creates a set of resources to a context, via creation or patching.
+   *
+   * @param context a context
+   * @param manifests a set of Kubernetes spec manifests
+   * @param action 'create' (only create new resources, do not overwrite existing) or 'apply' (create or patch as necessary)
+   * @param namespace the namespace to use for any resources that don't include one
+   * @return an array of resources created
+   *
+   * Heavily influenced by the API example:
+   * https://github.com/kubernetes-client/javascript/blob/0fbfd8fc2dcc7f4ec3e6fcd64a5c55169b6ef0b8/examples/typescript/apply/apply-example.ts
+   */
+  async syncResources(
+    context: string,
+    manifests: unknown[],
+    action: 'create' | 'apply',
+    namespace?: string,
+  ): Promise<KubernetesObject[]> {
+    const telemetryOptions: Record<string, unknown> = {
+      manifestsSize: manifests?.length,
+      action: action,
+    };
+    if (namespace) {
+      telemetryOptions.namespace = namespace;
+    }
+
     try {
       const ctx = new KubeConfig();
       ctx.loadFromFile(this.kubeconfigPath);
       ctx.currentContext = context;
-      for (const manifest of manifests) {
-        // https://github.com/kubernetes-client/javascript/issues/487
-        if (manifest?.metadata?.creationTimestamp) {
-          manifest.metadata.creationTimestamp = new Date(manifest.metadata.creationTimestamp);
+
+      const validSpecs = manifests.filter(s => (s as { kind: unknown })?.kind) as KubernetesObject[];
```

Comment:
I think the `(s as { kind: unknown })`should be moved to line 1095 where we defined manifests

if it's KubernetesObject then probably it's `manifests: KubernetesObject[]`

then you don't need to say s is from type {kind:unknown}

the ending cast should be a valid cast like refining a type but not goes straight from unknown to KubernetesObject (it's the manifests object that needs to be of KubernetesObject type)

Reply:
Agreed, fixed.